text/0567-use-and-resources.md

```diff
@@ -0,0 +1,1014 @@
+   `ember-concurrency` comes close, but only solves one specific problem.
+
+The `@use` API aims to solve both of these issues, and to do so in a way that
+integrates with autotracking so that users don't have to list out dependencies
+constantly in a repetitive manner. The core of the API consists of:
+
+- The `@use` decorator (and the complementary `{{use}}` helper)
+- The `Resource` base class (implemented via a `ResourceManager`, similar to
+  both components and modifiers).
+
+With these, we can create the `RemoteData` example at the beginning of this RFC
+like so:
+
+```js
+class RemoteData {
+  @tracked data = null;
+  @tracked isLoading = false;
+
+  get state() {
+    return {
+      isLoading: this.isLoading,
+      data: this.data,
+    };
+  }
+
+  async start(url) {
+    this.isLoading = true;
+
+    let response = await fetch(url);
+    let data = await response.json();
+
+    if (!this.isDestroyed) {
+      this.isLoading = false;
+      this.data = data;
+    }
+  }
+}
+
+export const remoteData = resource(RemoteData);
+```
+
+And we would use it like so:
+
+```js
+export default class SearchResults extends Component {
+  @use products = remoteData(this.args.searchUrl);
+}
+```
+
+```hbs
+{{#if this.products.isLoading}}
+  <LoadingSpinner/>
+{{else}}
+  {{#each this.products.data as |product|}}
+    <ProductCard @product={{product}} />
+  {{/each}}
+{{/if}}
+```
+
+### A Closer Look
+
+Resources fundamentally revolve around two lifecycle hooks: `start` and
+`teardown`.
+
+After a new instance of a `Resource` is created it runs the `start` method,
+which receives the arguments passed to `Resource.create()` and is used to
+setup the state. This method and its arguments are autotracked, and if anything
+that was consumed in it changes, `teardown` is called to destroy the
+`Resource`, and a new one is created with the new state in its place. When
+the _parent_ of the `Resource` (the class it is `@use`d on) is destroyed,
+then its `teardown` hook is called one last time and it is removed completely.
+
+Users can then define the `state` property, which is exposed externally on the
+field that was decorated with `@use`.
+
+```js
+class Counter {
+  @tracked count = 0;
+
+  intervalId = null;
+
+  // This is the value that is returned when `this.counter` is accessed on
+  // the component, either in the template, or in the JS of the component.
+  get state() {
+    return this.count;
+  }
+
+  // This runs the first time when an instance of MyComponent is created,
+  // and then whenever `this.args.interval` changes.
+  start(interval) {
+    this.intervalId = setInterval(() => this.count++, interval);
+  }
+
+  // This whenever `this.args.interval` changes, and when the parent
+  // MyComponent is destroyed.
+  teardown() {
+    clearInterval(this.intervalId);
+  }
+}
+
+const counter = resource(Counter);
+
+// usage
+class MyComponent extends Component {
+  @use counter = counter(this.args.interval);
```

Comment:
I'm finding this example a bit confusing because of the use of `counter` a few different ways. In the example above, you use different words on either side of the equals which I find easier to follow (`products = remoteData`). So, in a similar way, what if this was something like:

```suggestion
  @use counter = counterResource(this.args.interval);
```

This would also mean changing the variable on 243 to `const counterResource = resource(Counter)`